# Import Libraries and Dataset

In [1]:
# Import Libraries
import pandas as pd
import numpy as np
import os

In [2]:
# Create Path
path = "/Users/charlottelin/Documents/Achievement 6 Urban Flood Risk"

In [3]:
# Define path to dataset
file_path = os.path.join(path, '02 Data', 'Original Data', 'urban_pluvial_flood_risk_dataset.csv')

In [4]:
# Load dataset
df = pd.read_csv(file_path)

In [5]:
# Check import
df.head()

segment_id             city_name    admin_ward   latitude   longitude  \
0  SEG-00001    Colombo, Sri Lanka  Borough East   6.920633   79.912600   
1  SEG-00002        Chennai, India        Ward D  13.076487   80.281774   
2  SEG-00003      Ahmedabad, India     Sector 12  23.019473   72.638578   
3  SEG-00004      Hong Kong, China     Sector 14  22.302602  114.078673   
4  SEG-00005  Durban, South Africa      Sector 5 -29.887602   30.911008   

  catchment_id  elevation_m            dem_source       land_use soil_group  \
0      CAT-136          NaN  Copernicus_EEA-10_v5  Institutional        NaN   
1      CAT-049        -2.19  Copernicus_EEA-10_v5    Residential          D   
2      CAT-023        30.88             SRTM_3arc     Industrial          B   
3      CAT-168        24.28             SRTM_3arc    Residential          B   
4      CAT-171        35.70             SRTM_3arc     Industrial          C   

   drainage_density_km_per_km2  storm_drain_proximity_m storm_drain_type  \
0                         4.27                    160.5        CurbInlet   
1                         7.54                      NaN      OpenChannel   
2                        11.00                    152.5      OpenChannel   
3                         7.32                     37.0          Manhole   
4                         4.50                    292.4      OpenChannel   

  rainfall_source  historical_rainfall_intensity_mm_hr  return_period_years  \
0            ERA5                                 39.4                   50   
1            ERA5                                 56.8                   25   
2             IMD                                 16.3                    5   
3            ERA5                                 77.0                   10   
4            ERA5                                 20.8                    5   

                                  risk_labels  
0                                     monitor  
1  ponding_hotspot|low_lying|event_2025-05-02  
2                                     monitor  
3                                     monitor  
4                                     monitor

In [6]:
# Check data types
print (df.dtypes)

segment_id                              object
city_name                               object
admin_ward                              object
latitude                               float64
longitude                              float64
catchment_id                            object
elevation_m                            float64
dem_source                              object
land_use                                object
soil_group                              object
drainage_density_km_per_km2            float64
storm_drain_proximity_m                float64
storm_drain_type                        object
rainfall_source                         object
historical_rainfall_intensity_mm_hr    float64
return_period_years                      int64
risk_labels                             object
dtype: object


In [7]:
# check data numbers
print (df.shape)

(2963, 17)


In [8]:
# check other info about the dataset
print (df.info)

<bound method DataFrame.info of      segment_id             city_name    admin_ward   latitude   longitude  \
0     SEG-00001    Colombo, Sri Lanka  Borough East   6.920633   79.912600   
1     SEG-00002        Chennai, India        Ward D  13.076487   80.281774   
2     SEG-00003      Ahmedabad, India     Sector 12  23.019473   72.638578   
3     SEG-00004      Hong Kong, China     Sector 14  22.302602  114.078673   
4     SEG-00005  Durban, South Africa      Sector 5 -29.887602   30.911008   
...         ...                   ...           ...        ...         ...   
2958  SEG-02959         Paris, France        Ward B  48.872870    2.246250   
2959  SEG-02960       Shanghai, China     Sector 17  31.195529  121.435540   
2960  SEG-02961     Vancouver, Canada     Sector 12  49.162783 -123.037084   
2961  SEG-02962        Lagos, Nigeria        Zone V   6.504570    3.388571   
2962  SEG-02963          Osaka, Japan       Zone IV  34.740562  135.469263   

     catchment_id  elevation_m 

In [10]:
df.describe()

latitude    longitude  elevation_m  drainage_density_km_per_km2  \
count  2963.000000  2963.000000  2802.000000                  2679.000000   
mean     19.399103    31.676123    37.689818                     6.290866   
std      24.446320    79.542482    38.708958                     2.187365   
min     -36.999038  -123.292949    -3.000000                     1.270000   
25%       6.580006   -43.124581     8.725000                     4.670000   
50%      23.760802    36.889173    25.130000                     6.250000   
75%      37.887422   101.701136    59.620000                     7.830000   
max      55.821219   174.911271   266.700000                    12.070000   

       storm_drain_proximity_m  historical_rainfall_intensity_mm_hr  \
count              2724.000000                          2963.000000   
mean                123.202203                            43.811441   
std                 107.764541                            25.221200   
min                   0.200000                             5.400000   
25%                  47.975000                            25.800000   
50%                  91.700000                            37.900000   
75%                 162.625000                            55.550000   
max                 751.700000                           150.000000   

       return_period_years  
count          2963.000000  
mean             19.730003  
std              25.182883  
min               2.000000  
25%               5.000000  
50%              10.000000  
75%              25.000000  
max             100.000000

# Data Cleaning

## Identifying Missing Value

In [11]:
# Check for missing value
missing_values = df.isna().sum()

In [12]:
# Show columns with missing values
missing_values[missing_values > 0]

elevation_m                    161
soil_group                     362
drainage_density_km_per_km2    284
storm_drain_proximity_m        239
storm_drain_type               178
rainfall_source                315
dtype: int64

In [13]:
# Remove rows with 4 or more missing values
missing_per_row = df.isna().sum(axis=1)
df = df[missing_per_row < 4]

In [14]:
# Check new row count
print(df.shape)

(2962, 17)


#### This shows 1 row contained 4 or more missing values and was deleted. This shows that the dataset was relatively clean to begin with.

## Imputing Missing Values

### Elevation

In [15]:
# Replace missing elevation values by median
df['elevation_m'] = df.groupby('city_name')['elevation_m'].transform(lambda x: x.fillna(x.median()))

### Soil Group

In [16]:
# Replace missing soil group value by mode
df['soil_group'].mode()

0    B
Name: soil_group, dtype: object

In [17]:
df['soil_group'] = df['soil_group'].fillna('B')

### Drainage Density

In [18]:
# Replace drainage density values within land_use groups using mean
df['drainage_density_km_per_km2'] = df.groupby('land_use')['drainage_density_km_per_km2'].transform(
    lambda x: x.fillna(x.mean()))

### Storm Drain Proximity

In [19]:
# Replace storm drain proximity within land_use groups using mean
df['storm_drain_proximity_m'] = df.groupby('land_use')['storm_drain_proximity_m'].transform(
    lambda x: x.fillna(x.mean()))

### Storm Drain Type

In [20]:
# Check storm drain types
df['storm_drain_type'].value_counts()

storm_drain_type
CurbInlet      842
Manhole        749
GratedInlet    644
OpenChannel    549
Name: count, dtype: int64

In [21]:
# Replace missing storm drain types with "Unknown"
df['storm_drain_type'] = df['storm_drain_type'].fillna('Unknown')

### Rainfall Source

In [22]:
# Check rainfall source types
df['rainfall_source'].value_counts()

rainfall_source
ERA5          1334
LocalGauge     502
Blended        445
IMD            367
Name: count, dtype: int64

In [23]:
# Replace missing rainfall source with "Unknown"
df['rainfall_source'] = df['rainfall_source'].fillna('Unknown')

### Final Missing Value Check

In [24]:
df.isnull().sum()

segment_id                             0
city_name                              0
admin_ward                             0
latitude                               0
longitude                              0
catchment_id                           0
elevation_m                            0
dem_source                             0
land_use                               0
soil_group                             0
drainage_density_km_per_km2            0
storm_drain_proximity_m                0
storm_drain_type                       0
rainfall_source                        0
historical_rainfall_intensity_mm_hr    0
return_period_years                    0
risk_labels                            0
dtype: int64

#### Data cleaning complete

## Basic Exploratory Data Analysis

In [25]:
# Review variables
profile = pd.DataFrame({
    "dtype": df.dtypes,
    "n_unique": df.nunique(),
    "n_missing": df.isna().sum(),
    "pct_missing": df.isna().mean().round(3) * 100
})
profile

dtype  n_unique  n_missing  pct_missing
segment_id                            object      2962          0          0.0
city_name                             object        63          0          0.0
admin_ward                            object        91          0          0.0
latitude                             float64      2962          0          0.0
longitude                            float64      2962          0          0.0
catchment_id                          object       180          0          0.0
elevation_m                          float64      2229          0          0.0
dem_source                            object         5          0          0.0
land_use                              object         9          0          0.0
soil_group                            object         4          0          0.0
drainage_density_km_per_km2          float64       855          0          0.0
storm_drain_proximity_m              float64      1750          0          0.0
storm_drain_type                      object         5          0          0.0
rainfall_source                       object         5          0          0.0
historical_rainfall_intensity_mm_hr  float64       866          0          0.0
return_period_years                    int64         6          0          0.0
risk_labels                           object       195          0          0.0

In [29]:
# Descriptive Stats for the cleaned dataset
df.describe()

latitude    longitude  elevation_m  drainage_density_km_per_km2  \
count  2962.000000  2962.000000  2962.000000                  2962.000000   
mean     19.392549    31.712840    37.594541                     6.292750   
std      24.447844    79.530794    37.999568                     2.110028   
min     -36.999038  -123.292949    -3.000000                     1.270000   
25%       6.579197   -43.120010     9.240000                     4.752500   
50%      23.760790    36.890411    26.170000                     6.280000   
75%      37.886114   101.701955    58.047500                     7.670000   
max      55.821219   174.911271   266.700000                    12.070000   

       storm_drain_proximity_m  historical_rainfall_intensity_mm_hr  \
count              2962.000000                          2962.000000   
mean                123.072485                            43.815057   
std                 103.819179                            25.224690   
min                   0.200000                             5.400000   
25%                  51.125000                            25.800000   
50%                 100.450000                            37.900000   
75%                 160.650000                            55.575000   
max                 751.700000                           150.000000   

       return_period_years  
count          2962.000000  
mean             19.734976  
std              25.185680  
min               2.000000  
25%               5.000000  
50%              10.000000  
75%              25.000000  
max             100.000000

## Summary of Data Cleaning & EDA

The dataset has been cleaned and is now analysis-ready:
- Missing values were addressed using contextual imputation strategies:
  - `drainage_density_km_per_km2` and `storm_drain_proximity_m` filled by land use group mean.  
  - `storm_drain_type` and `rainfall_source` imputed with `"Unknown"`.    
- All categorical variables were standardized for consistency.  
  
Basic descriptive analysis provided a first look at the dataset:
- Elevation values range from -3 m (sentinel) to ~267 m, with most segments below 50 m.  
- Drainage density averages ~6.3 km/km², while storm drain proximity averages ~123 m.  
- The most common storm drain types are Curb Inlet (28%) and Manhole (25%).  
- Rainfall sources are led by ERA5 (45%), followed by local gauges (17%).  

Overall, the dataset was relatively clean to begin with, and is now well-structured, complete, and ready for further analysis and visualization.

# Export Cleaned Dataset

In [30]:
# Export the cleaned csv to the 'Prepared Data' folder
df.to_csv(
    os.path.join(path, '02 Data', 'Prepared Data', 'urban_pluvial_flood_risk_clean.csv'),
    index=False)